Add hitsound checks

Naxesss opened the pull request:
Makes further progress towards resolving https://github.com/ppy/osu/issues/12091.

Adds
- `CheckMutedObjects`
    - Ensures clickable objects have audible feedback.
    - Also adds a negligible check for muted non-clickable edges (e.g. slider repeats/tails).
- `CheckFewHitsounds`
    - Ensures that beatmaps are hitsounded (claps, whistles, finishes) and don't have long periods of no hitsounds.
    - Only points out periods in which there could be hitsounds added (i.e. where there are hitnormals).
- Tests for those checks

https://user-images.githubusercontent.com/30292137/123521124-f81c6580-d6b4-11eb-9d97-287026e0d03f.mp4

---
These were difficult to test and show in-game due to hitsound controls being lacking; mainly the ability to add/remove samples to/from nested hitobjects like slider heads. The `MockNestableHitObject` class (name could probably be better) in `CheckMutedObjectsTest` and its respective tests hopefully help make sure that works properly.

Naxesss committed:
Add muted objects check

Naxesss committed:
Add few hitsounds check

Naxesss committed:
Add muted object check tests

Naxesss committed:
Add few hitsounds check tests

Naxesss committed:
Add hitsounded with break test

Naxesss committed:
Make || and && priority explicit

Naxesss committed:
Factor out edge type logic

bdach commented on osu.Game.Tests/Editing/Checks/CheckFewHitsoundsTest.cs:
```diff
@@ -0,0 +1,191 @@
+                hitObjects.Add(new HitCircle { StartTime = 1000 * i, Samples = samples });
+            }
+
+            // Should prompt one warning between 1st and 11th, and another between 11th and 20th.
```
I don't think that's right? The 16th is hitsounded so it technically breaks a period of silence. On debugging the even the check result doesn't agree, it reports one warning between 1st and 11th, and another between 16th and 30th.

Naxesss replied:
Oops forgot to update the comment

bdach commented on osu.Game.Tests/Editing/Checks/CheckFewHitsoundsTest.cs:
```diff
@@ -0,0 +1,191 @@
+                hitObjects.Add(new HitCircle { StartTime = 1000 * i, Samples = samples });
+            }
+
+            // Should prompt one problem between 1st and 40th, and another between 40th and 80th.
```
Nit but this is an off-by-one; `i` is zero-based, so the `i == 40` check above actually modifies the 41st.

bdach commented on osu.Game.Tests/Editing/Checks/CheckFewHitsoundsTest.cs:
```diff
@@ -0,0 +1,191 @@
+                hitObjects.Add(new HitCircle { StartTime = 1000 * i, Samples = samples });
+            }
+
+            // Should prompt one problem between 1st and 40th, and another between 40th and 80th.
```
copy-pasted once too much

bdach commented on osu.Game.Tests/Editing/Checks/CheckFewHitsoundsTest.cs:
```diff
@@ -0,0 +1,191 @@
+namespace osu.Game.Tests.Editing.Checks
+{
+    [TestFixture]
+    public class CheckFewHitsoundsTest
```
As far as I can see the nested HO case has no test coverage; would like to see some if possible.

Naxesss replied:
`MockNestableHitObject` could probably be reused for that. I'm not entirely sure where to put it though. Any ideas? Also do mock classes generally go in the test or non-test projects?

bdach replied:
I think putting it in `osu.Game.Tests.Editing.Checks` for now might be ok. Could see it being in `osu.Game.Tests.Beatmaps` as well.

bdach commented on osu.Game/Rulesets/Edit/Checks/CheckFewHitsounds.cs:
```diff
@@ -0,0 +1,161 @@
+        private const int warning_threshold_objects = 4;
+        private const int problem_threshold_objects = 16;
+
+        private static readonly string[] hitsound_types = { HitSampleInfo.HIT_CLAP, HitSampleInfo.HIT_WHISTLE, HitSampleInfo.HIT_FINISH };
```
Can be replaced with `HitSampleInfo.AllAdditions`.

bdach commented on osu.Game/Rulesets/Edit/Checks/CheckFewHitsounds.cs:
```diff
@@ -0,0 +1,161 @@
+
+        public CheckMetadata Metadata { get; } = new CheckMetadata(CheckCategory.Audio, "Few or no hitsounds");
+
+        public IEnumerable<IssueTemplate> PossibleTemplates => new IssueTemplate[]
+        {
+            new IssueTemplateLongPeriodProblem(this),
+            new IssueTemplateLongPeriodWarning(this),
+            new IssueTemplateNoHitsounds(this)
+        };
```
`LongPeriodNegligible` is not listed. Not sure whether that was intentional, because as far as I see other check types also add negligibles here (`CheckUnsnappedObjects` does at least).

Naxesss replied:
Good catch

bdach commented on osu.Game/Rulesets/Edit/Checks/CheckFewHitsounds.cs:
```diff
@@ -0,0 +1,161 @@
+            new IssueTemplateNoHitsounds(this)
+        };
+
+        private bool hasHitsounds;
```
Would rename to `mapHasHitsounds` to clarify intent. As is the `hasHitsound(hitObject) || (isLastObject && hasHitsounds)` check lower down reads a touch weird.

bdach commented on osu.Game/Rulesets/Edit/Checks/CheckFewHitsounds.cs:
```diff
@@ -0,0 +1,161 @@
+                    yield return new IssueTemplateLongPeriodNegligible(this).Create(lastHitsoundTime, timeWithoutHitsounds);
+            }
+
+            if (hasHitsound(hitObject))
```
Called the second time in this method. If `hasHitsounds` was to be renamed as suggested, the result of `hasHitsound()` could be stored to a local for a bit less of wastefulness.

bdach commented on osu.Game/Rulesets/Edit/Checks/CheckFewHitsounds.cs:
```diff
@@ -0,0 +1,161 @@
+
+            for (int i = 0; i < hitObjectCount; ++i)
+            {
+                var hitObject = context.Beatmap.HitObjects[i];
```
Side note but the way this algorithm is set up (iterating topmost objects in the outer loop and nested objects within the inner) is relying on two assumptions, namely:

* There are no overlapping objects / objects are sorted by their end time
* Nested hit objects are sorted by their end time, and their end time fits within the parent object's duration

Those are generally *fine* for the typical case but I'm afraid that it may break down a bit in cases where concurrent objects become a possibility (for osu! this isn't much of a problem, but for mania it sort of can become one). It may be better to collect all hitobjects and nested hitobjects to a list and then order them by end time explicitly to remove the aforementioned assumptions.

Naxesss replied:
That is a case I hadn't considered. After fixing it like you suggested my initial thought was to add a test, but thinking of a scenario in which this would be an issue is quite difficult without getting too specific. So am thinking I skip testing this case.

bdach replied:
Here's a test case that would fail with the old code but passes with the new.

```csharp
[Test]
public void TestConcurrentObjects()
{
    var hitObjects = new List<HitObject>();

    var notHitsounded = new List<HitSampleInfo> { new HitSampleInfo(HitSampleInfo.HIT_NORMAL) };
    var hitsounded = new List<HitSampleInfo>
    {
        new HitSampleInfo(HitSampleInfo.HIT_NORMAL),
        new HitSampleInfo(HitSampleInfo.HIT_FINISH)
    };

    var ticks = new List<HitObject>();
    for (int i = 1; i < 10; ++i)
        ticks.Add(new SliderTick { StartTime = 5000 * i, Samples = hitsounded });

    var nested = new MockNestableHitObject(ticks.ToList(), 0, 25000)
    {
        Samples = notHitsounded
    };
    nested.ApplyDefaults(new ControlPointInfo(), new BeatmapDifficulty());
    hitObjects.Add(nested);

    for (int i = 1; i < 10; ++i)
        hitObjects.Add(new HitCircle { StartTime = 10000 * i, Samples = notHitsounded });

    assertOk(hitObjects);
}
```

I'm not necessarily sure that this is permitted from a RC/mapping good practice standpoint, it's just a test case that would fail before due to not considering the possibility of nested HOs concurrent with normal HOs.

Naxesss replied:
I can't reproduce the failure. Reverting https://github.com/ppy/osu/pull/13669/commits/0c0fd291d9381d718668e0588d6fb5dee60e91c7 and testing this results in success. So does using the exact code from [the initial commit of `CheckFewHitsounds`](https://github.com/ppy/osu/blob/4b436b774dcb3c35d145410fb56edc12e2c66ebb/osu.Game/Rulesets/Edit/Checks/CheckFewHitsounds.cs).

In terms of the case being realistic/practical, it would be if we replace the slider ticks with repeats (ticks can't have hitnormals or additions, but repeats can). So it's fine in that regard.

bdach replied:
I had actually tested that it failed on the original rev, but then edited some things and forgot to re-test again. I've updated the comment above - should now fail before the change. There were not enough hitcircles to trigger a warning (didn't cross the 4 object threshold).

bdach commented on osu.Game/Rulesets/Edit/Checks/CheckMutedObjects.cs:
```diff
@@ -0,0 +1,147 @@
+{
+    public class CheckMutedObjects : ICheck
+    {
+        /// <summary>
+        /// Volume percentages lower than this are typically inaudible.
+        /// </summary>
+        private const int muted_threshold = 5;
```
Nit: xmldoc is a touch misleading, given that the conditionals that use this do not use strict equality (so this should read "lower than or equal to" for 100% correctness).

bdach commented on osu.Game/Rulesets/Edit/Checks/CheckMutedObjects.cs:
```diff
@@ -0,0 +1,147 @@
+            if (hitObject is IHasRepeats hasRepeats)
+            {
+                double spanDuration = hasRepeats.Duration / hasRepeats.SpanCount();
+                if (Precision.AlmostEquals((time - hitObject.StartTime) % spanDuration, 0f, 1f))
```
I generally do not recommend using floating-point modulo because it can have unexpected properties:

```
12.1 % 0.1 = 0.09999999999999898
 0.2 % 0.1 = 0
-0.1 % 0.1 = -0
-0.3 % 0.1 = -0.09999999999999998
```

In this case there's nothing that guarantees you won't get a value just ever so slightly below `spanDuration`. I'd recommend using `(time - hitObject.StartTime) / spanDuration` instead and probably doing iterative `AlmostEquals()` checks with 1, 2, 3, ... until that division result is exceeded.

Naxesss replied:
Ended up checking if the result of the division is a whole number:
```cs
double spans = (time - hitObject.StartTime) / spanDuration;

if (Precision.AlmostEquals(spans, Math.Ceiling(spans)) || Precision.AlmostEquals(spans, Math.Floor(spans)))
    return EdgeType.Repeat;
```
Should accomplish the same thing without iteration. The acceptable difference on `AlmostEquals` may be a bit too small as-is though, would depend on `spanDuration`.

bdach replied:
I think that for something like this a mix of relative and absolute error would be the absolute best (something like `Math.Max(1e-3, 1 / spanDuration)`) when it comes to the precision, but for simplicity going with something constant like `1e-3` might be good enough. Not sure I'd leave the currently-used default of `1e-7` as it seems a bit on the smaller side.

Naxesss replied:
Ended up going for `1 / spanDuration` (1 ms) for consistency with the acceptable difference of head/tail edges. Likelihood that `spanDuration` is so large that the resulting acceptable difference is less than E-7 is quite low in any realistic case, I think.

bdach replied:
Sure, I'd say that also makes sense.

Naxesss committed:
Fix comments in few hitsounds check tests

Naxesss committed:
Use `HitSampleInfo.AllAdditions` instead of new list

Naxesss committed:
Add negligible template to `PossibleTemplates`

Naxesss committed:
Rename `hasHitsounds` -> `mapHasHitsounds`

Naxesss committed:
Use local variables for `hasHitsound` & `couldHaveHitsound`

Naxesss committed:
Order hitobjects by endtime

Naxesss committed:
Add "or equal to" to volume threshold xmldocs

Naxesss committed:
Improve precision for repeat edges

Naxesss committed:
Move `MockNestedHitObject` to own class

Naxesss committed:
Add concurrent hitobjects test for few hitsounds check

See https://github.com/ppy/osu/pull/13669#discussion_r659314980

Naxesss committed:
Improve acceptable difference for repeat edges

Likelihood that `spanDuration` is greater than E+7 is quite low in any realistic case, so this should work fine.

Naxesss committed:
Add check for `spanDuration` <= 0 prior to division

Naxesss committed:
Use local variables for common sample lists

Naxesss committed:
Add nested hitobject tests for few hitsounds check

peppy reviewed (approved):
Solid as always 👍🏻
